File: backend/app/translation/engine.py
"""Orchestrates config resolution, harness rendering, and the provider call.

The engine is the seam the routes depend on. It knows nothing about HTTP and
nothing about OpenRouter specifically — it composes a ``ConfigStore`` and any
``ChatProvider``. That makes it trivial to test against a fake provider and to
reuse from a CLI or a queue worker later.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import AsyncIterator

from app.config_store import ConfigStore
from app.harness import build_chat_request, extract_translation
from app.harness.models import TranslationSpec
from app.providers.base import ChatProvider, Credentials, Usage
from app.schemas import TranslateRequest


@dataclass(slots=True)
class TranslationOutcome:
    translation: str
    spec: TranslationSpec
    model: str
    usage: Usage | None


class TranslationEngine:
    def __init__(self, store: ConfigStore, provider: ChatProvider) -> None:
        self._store = store
        self._provider = provider

    async def translate(
        self, req: TranslateRequest, credentials: Credentials
    ) -> TranslationOutcome:
        spec = self._store.resolve(req)
        chat_request = build_chat_request(spec)
        result = await self._provider.complete(chat_request, credentials)
        translation = extract_translation(result.text, spec)
        return TranslationOutcome(
            translation=translation,
            spec=spec,
            model=result.model,
            usage=result.usage,
        )

    async def stream(
        self, req: TranslateRequest, credentials: Credentials
    ) -> AsyncIterator[str]:
        """Stream raw output deltas.

        Delimiter-based extraction is intentionally *not* applied here: it needs
        the full reply. Profiles intended for streaming should set empty
        delimiters so the raw deltas are already clean.
        """
        spec = self._store.resolve(req, stream=True)
        chat_request = build_chat_request(spec)
        async for delta in self._provider.stream(chat_request, credentials):
            yield delta

    @property
    def store(self) -> ConfigStore:
        return self._store

File: worker/src/engine.ts
// Orchestrates config resolution, harness rendering, and the provider call —
// the TS mirror of translation/engine.py. Knows nothing about HTTP or OpenRouter
// specifically; composes a ConfigStore and any ChatProvider.

import type { ConfigStore } from "./config";
import { buildChatRequest, extractTranslation } from "./harness";
import type { TranslationSpec } from "./harness";
import type { TranslateRequest } from "./schemas";
import type { ChatProvider, Credentials, Usage } from "./types";

export interface TranslationOutcome {
  translation: string;
  spec: TranslationSpec;
  model: string;
  usage?: Usage;
}

export class TranslationEngine {
  constructor(
    private readonly store: ConfigStore,
    private readonly provider: ChatProvider,
  ) {}

  async translate(
    req: TranslateRequest,
    credentials: Credentials,
  ): Promise<TranslationOutcome> {
    const spec = this.store.resolve(req);
    const chatRequest = buildChatRequest(spec);
    const result = await this.provider.complete(chatRequest, credentials);
    return {
      translation: extractTranslation(result.text, spec),
      spec,
      model: result.model,
      usage: result.usage,
    };
  }

  // Streams raw output deltas. Delimiter extraction is intentionally not applied
  // (it needs the full reply); streaming resolves to a delimiter-free profile.
  async *stream(
    req: TranslateRequest,
    credentials: Credentials,
  ): AsyncIterable<string> {
    const spec = this.store.resolve(req, { stream: true });
    const chatRequest = buildChatRequest(spec);
    yield* this.provider.stream(chatRequest, credentials);
  }
}
